lib/promoted/ruby/client/validator.rb

```diff
@@ -2,26 +2,151 @@ module Promoted
+                    if field[:required] then
+                        raise ValidationError.new(field[:name].to_s + " is required on " + obj_name) if !obj.has_key?(field[:name])
+                    end
+
+                    if field[:type] && obj.has_key?(field[:name]) then
+                        raise ValidationError.new(field[:name].to_s + " should be a " + field[:type].to_s) if !obj[field[:name]].is_a?(field[:type])
+                    end
                 }
             end
 
-            def self.validate_metrics_request!(metrics_req)
-                self.validate_fields!(
+            def validate_user_info!(ui)
+                validate_fields!(
+                    ui,
+                    "user info",
+                    [
+                        {
+                            :name => :user_id,
+                            :type => String
+                        },
+                        {
+                            :name => :log_user_id,
+                            :type => String
+                        }
+                    ]
+                )
+            end
+
+            def validate_insertion!(ins)
+                validate_fields!(
+                    ins,
+                    "insertion",
+                    [
+                        {
+                            :name => :platform_id,
+                            :type => Integer
+                        },
+                        {
+                            :name => :insertion_id,
+                            :type => String
+                        },
+                        {
+                            :name => :request_id,
+                            :type => String
+                        },
+                        {
+                            :name => :view_id,
+                            :type => String
+                        },
+                        {
+                            :name => :session_id,
+                            :type => String
+                        },
+                        {
+                            :name => :content_id,
+                            :type => String
+                        },
+                        {
+                            :name => :position,
+                            :type => Integer
+                        },
+                        {
+                            :name => :delivery_score,
+                            :type => Integer
+                        }
+                    ]
+                )
+
+                if ins[:user_info] then
+                    self.validate_user_info! ins[:user_info]
+                end
+            end
+
+            def validate_request!(req)
+                validate_fields!(
+                    req,
+                    "request",
+                    [
+                        {
+                            :name => :platform_id,
```

Comment:
side-comment - Right now, we're not forcing clients to set this.  We're just doing it on the server-side.  It's fine to leave it in.  We'll eventually ask for it.

Reply:
Ack. I was mostly trying to mimic the TypeScript annotations' behaviors with the field validators.